// tronweb/lib/commonjs/utils/help.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.stringUtf8toHex = exports.hexString2Utf8 = exports.hexString2Address = exports.address2HexString = exports.stringUtf8tHex = exports.hexStringToUtf8 = exports.base58ToHexString = exports.hexStringToBase58 = void 0;
const code_js_1 = require("./code.js");
const address_js_1 = require("./address.js");
const crypto_js_1 = require("./crypto.js");
const bytes_js_1 = require("./bytes.js");
function hexStringToBase58(sHexString) {
    if (sHexString.length < 2 || (sHexString.length & 1) != 0)
        return '';
    const bytes = (0, code_js_1.hexStr2byteArray)(sHexString);
    return (0, crypto_js_1.getBase58CheckAddress)(bytes);
}
exports.hexStringToBase58 = hexStringToBase58;
function base58ToHexString(sBase58) {
    const bytes = (0, crypto_js_1.decodeBase58Address)(sBase58);
    if (!bytes)
        return '';
    return (0, bytes_js_1.byteArray2hexStr)(bytes);
}
exports.base58ToHexString = base58ToHexString;
function hexStringToUtf8(hex) {
    const arr = hex.split('');
    let out = '';
    for (let i = 0; i < arr.length / 2; i++) {
        const tmp = `0x${arr[i * 2]}${arr[i * 2 + 1]}`;
        const charValue = String.fromCharCode(parseInt(tmp));
        out += charValue;
    }
    return out;
}
exports.hexStringToUtf8 = hexStringToUtf8;
function stringUtf8tHex(str) {
    let val = '';
    for (let i = 0; i < str.length; i++) {
        if (val == '')
            val = str.charCodeAt(i).toString(16);
        else
            val += str.charCodeAt(i).toString(16);
    }
    return val;
}
exports.stringUtf8tHex = stringUtf8tHex;
function address2HexString(sHexAddress) {
    if (sHexAddress.length == 42 && sHexAddress.indexOf(address_js_1.ADDRESS_PREFIX) == 0)
        return sHexAddress;
    return base58ToHexString(sHexAddress);
}
exports.address2HexString = address2HexString;
function hexString2Address(sAddress) {
    return hexStringToBase58(sAddress);
}
exports.hexString2Address = hexString2Address;
function hexString2Utf8(sHexString) {
    return hexStringToUtf8(sHexString);
}
exports.hexString2Utf8 = hexString2Utf8;
function stringUtf8toHex(sUtf8) {
    return stringUtf8tHex(sUtf8);
}
exports.stringUtf8toHex = stringUtf8toHex;
